mk/patterns.mk

```diff
@@ -1,11 +1,31 @@
+
+# These are the complete command lines we use to compile C and C++ files. We separate them out so
+# we can use them for both compiling and creaating the compile_commands.json file.
+
+CC_CMD=$(CC) -o $@ -c $< $(CPPFLAGS) $(GLOBAL_CFLAGS) $(CFLAGS) $($@_CFLAGS) -MMD -MF $(call filename-to-dep, $@) -MP
+
+CXX_CMD=$(CXX) -o $@ -c $< $(CPPFLAGS) $(GLOBAL_CXXFLAGS_PCH) $(GLOBAL_CXXFLAGS) $(CXXFLAGS) $($@_CXXFLAGS) $(ERROR_SWITCH_ENUM) -MMD -MF $(call filename-to-dep,$@) -MP
```

Comment:
(nit, don't do if that turns out to be too complex): Could this be made a [function](https://www.gnu.org/software/make/manual/html_node/Call-Function.html) instead to avoid the (hard to understand) `let` calls below?

Reply:
Done, I also added comments.